### ISSUE TITLE
AWS::EC2::VPCEndpoint return ENI IP address

### ISSUE THREAD

**Issue by hoegertn** [aws-cloudformation/aws-cloudformation-coverage-roadmap#109]
For interface typed VPC endpoints it would be good to have the private IPs of the ENIs as an attribute accessible via !GetAtt.

For several use cases, DNS names are not enough and you need the IP of the ENI.

Currently, this can only be solved by a custom resource that takes the ENIs (`!GetAtt VPCE.NetworkInterfaceIds`) and resolves them to IPs.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-vpcendpoint.html

Category: Networking & Content (VPC, Route53, API GW,...)

**Comment by sc-pulsion** [comment]
@hoegertn I dont suppose you have an example of that custom resource to save me re-inventing the wheel? Thanks.

**Comment by moqrin3** [comment]
https://dev.classmethod.jp/cloud/aws/transfer-for-sftp-restrict-ip-cloudformation/

**Comment by dkavanagh** [comment]
Yes, being able to get these IPs easily and throw them in an NLB TargetGroup would be very handy. Off to write a custom resource to look them up. sigh

**Comment by hunyhabib** [comment]
I am also facing the same issue, I am trying to create a IP Target group to include VPC endpoint ENI IPs using cloud formation.  with the GetAtt i could get the DNS but as mentioned above there is some scenarios we need the IPs.

`VPCEndpoint:
    Type: AWS::EC2::VPCEndpoint
    Properties: 
      PrivateDnsEnabled: true
      VpcEndpointType: Interface
      SecurityGroupIds: 
        - !ImportValue 
          'Fn::Sub': '${InfrastructureStackName}-LoadBalancerSecuritGroup'
      ServiceName: !Sub 'com.amazonaws.${AWS::Region}.execute-api'
      SubnetIds: !Ref Subnets
      VpcId: !Ref VPC
  TargetGroup:
    Type: 'AWS::ElasticLoadBalancingV2::TargetGroup'
    Properties:
      HealthCheckIntervalSeconds: 300
      HealthCheckPath: /index.html
      HealthCheckTimeoutSeconds: 5
      UnhealthyThresholdCount: 2
      HealthyThresholdCount: 2
      Name: !Sub 'targetgroup-${EnvironmentName}'
      Port: !Ref TargetPort
      Protocol: HTTP
      TargetGroupAttributes:
        - Key: deregistration_delay.timeout_seconds
          Value: 60
      TargetType: ip <------------------ I want to include the IPs
      VpcId: !Ref VPC `

**Comment by anubhav-jhalani** [comment]
> For interface typed VPC endpoints it would be good to have the private IPs of the ENIs as an attribute accessible via !GetAtt.
> 
> For several use cases, DNS names are not enough and you need the IP of the ENI.
> 
> Currently, this can only be solved by a custom resource that takes the ENIs (`!GetAtt VPCE.NetworkInterfaceIds`) and resolves them to IPs.
> 
> https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-vpcendpoint.html
> 
> Category: Networking & Content (VPC, Route53, API GW,...)

@hoegertn  Could you please specify that custom resource? I am also looking for a solution of this problem

**Comment by zindello** [comment]
+1 I would also like to see this as I'm attaching an NLB in front of an IOT Core MQTT endpoint that's accessed over a direct connect (because reasons)

**Comment by jasonrichardsmith** [comment]
The diagram of your architecture here.
https://docs.aws.amazon.com/whitepapers/latest/best-practices-api-gateway-private-apis-integration/rest-api.html#private-integration
Would require a return of IP addresses. Otherwise it is done with a Custom Resource.

**Comment by dkasten79** [comment]
+1

**Comment by DonthineniSagar** [comment]
+1

**Comment by Landerstraeten** [comment]
+1

**Comment by chufgard** [comment]
Four years and no progress at all?
I want to connect a NLB to an API Gateway ( https://docs.aws.amazon.com/whitepapers/latest/best-practices-api-gateway-private-apis-integration/rest-api.html#private-integration ) and it seems that I must use a Custom Resource.

**Comment by muhamadto** [comment]
+1

**Comment by dalibor-petras** [comment]
+1 pls

**Comment by gjkraken** [comment]
+1 🙏

**Comment by mattyboy84** [comment]
+1

**Comment by agupta94** [comment]
+1
